util/intensity_explorer.py

```diff
@@ -0,0 +1,623 @@
+            # use the measured variance instead in this case.
+            if count == 1:
+              variances.append(rtable['intensity.sum.variance'][j-1])
+            else:
+              sum_weight = sum_weights[j-1]
+              var_weight = 1 / (sum_weight -
+                                sum_square_weights[j-1] / sum_weight)
+              variances.extend(
+                flex.double(count, weighted_sum_square_residual * var_weight))
+            # Reinitialise:
+            prev_index = index
+            count = 1
+            weighted_sum_square_residual = weight * residual * residual
+          # Handle the first row:
+          else:
+            prev_index = rtable['miller_index.asu'][j]
+            count = 1
+            weighted_sum_square_residual = weight * residual * residual
+        # Record the aggregated values for the last group:
+        # The weighted variance is undefined for multiplicity=1,
+        # use the measured variance instead in this case.
+        if count == 1:
+          variances.append(rtable['intensity.sum.variance'][-1])
+        else:
+          sum_weight = sum_weights[-1]
+          var_weight = 1 / (sum_weight - sum_square_weights[-1] / sum_weight)
+          variances.extend(
+              flex.double(count, weighted_sum_square_residual * var_weight))
+        # Record the variances in the reflection table.
+        rtable['intensity.mean.variance'] = variances
+
+      self.rtables[key] = rtable
+
+  def _make_z(self, uncertainty='sigma'):
+    u"""
+    Generate reflection z-scores.
+
+    Calculate z-scores from reflection intensities, weighted mean
+    intensities and a chosen measure of uncertainty in the intensity
+    measurement.
+
+    :param uncertainty: Chosen measure of uncertainty.  Options are
+      * ``stddev`` — standard deviation, as calculated from the unbiased
+      weighted variance aggregated amongst all symmetry-equivalent reflections;
+      * ``sigma`` — measurement error for individual reflections.
+    :type uncertainty: str
+    """
+
+    uncertainty_name = {
+      'sigma': 'intensity.sum.variance',
+      'stddev': 'intensity.mean.variance'
+    }[uncertainty]
+
+    for key, rtable in self.rtables.items():
+      try:
+        uncertainty_value = flex.sqrt(rtable[uncertainty_name])
+      except KeyError:
+        from logging import warn
+        uncertainty_value = flex.sqrt(rtable['intensity.sum.variance'])
+        warn(u"""Weighted variances haven't been calculated,
+      be sure to specify calculate_variances=True to use them.
+      Defaulting to measured σ values as a measure of uncertainty instead.""")
+
+      z = (rtable['intensity.sum.value']
+           - rtable['intensity.mean.value']) / uncertainty_value
+      rtable['intensity.z_score'] = z
+
+      self.rtables[key] = rtable
+
+  def _probplot_data(self):
+    """Generate the data for a normal probability plot of z-scores."""
+
+    from scipy import stats as ss
```

Comment:
I think ```import scipy.stats``` takes less effort to understand, even though the method is a short one.